### PR TITLE
Ensure that if grpc stream terminates that DirectStreamObserver does not block on Phaser.

### PR DESCRIPTION
**Pull request by scwhittle**
This can fix classes of stuckness during termination if the channel breaks but threads are blocked on the
phaser attempting to send.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by scwhittle** [comment]
Run Java PreCommit

**Comment by scwhittle** [comment]
assign to next reviewer

**Comment by scwhittle** [comment]
R: @Abacn

File: sdks/java/fn-execution/src/main/java/org/apache/beam/sdk/fn/stream/DirectStreamObserver.java [hunk 1]
```diff
@@ -101,6 +101,11 @@ public void onNext(T value) {
                     + "this issue.",
```
Comment (Abacn) [inline sdks/java/fn-execution/src/main/java/org/apache/beam/sdk/fn/stream/DirectStreamObserver.java:101]: "https://issues.apache.org/jira/browse/BEAM-4280 for the history for this issue."

shall we update this link (to a GitHub Issue) and also print in the warn message / or remove the outdated JIRA link at INFO message

File: sdks/java/fn-execution/src/main/java/org/apache/beam/sdk/fn/stream/ForwardingClientResponseObserver.java [hunk 2]
```diff
@@ -25,23 +25,27 @@
...
+ * the stream terminates.
  *
  * <p>This is as thread-safe as the underlying stream observer that is being wrapped.
  */
 public final class ForwardingClientResponseObserver<ReqT, RespT>
     implements ClientResponseObserver<RespT, ReqT> {
   public static <ReqT, RespT> ForwardingClientResponseObserver<ReqT, RespT> create(
-      StreamObserver<ReqT> inbound, Runnable onReadyHandler) {
-    return new ForwardingClientResponseObserver<>(inbound, onReadyHandler);
+      StreamObserver<ReqT> inbound, Runnable onReadyHandler, Runnable onDoneHandler) {
```
Comment (Abacn) [inline sdks/java/fn-execution/src/main/java/org/apache/beam/sdk/fn/stream/ForwardingClientResponseObserver.java:36]: public method signature changed. iiuc this is internal class, should it actually be package private; or keep the original overwrite method (without onDoneHandler)?

**Review by Abacn — COMMENTED** [review]
Thanks, this lgtm, just added a few optional comment

File: sdks/java/fn-execution/src/main/java/org/apache/beam/sdk/fn/stream/ForwardingClientResponseObserver.java (hunk 2 above)
Comment (scwhittle) [reply sdks/java/fn-execution/src/main/java/org/apache/beam/sdk/fn/stream/ForwardingClientResponseObserver.java:36]: It is internal AFAIK but it seems safer to add overload so went with that.

File: sdks/java/fn-execution/src/main/java/org/apache/beam/sdk/fn/stream/DirectStreamObserver.java (hunk 1 above)
Comment (scwhittle) [reply sdks/java/fn-execution/src/main/java/org/apache/beam/sdk/fn/stream/DirectStreamObserver.java:101]: I don't think it is worthwhile history or filing an issue. I just tried to update comment to summarize what was going on. PTAL

**Comment by Abacn** [comment]
github action failures due to #27854 not related

**Pull request merged**